### PR TITLE
Multiple endstops on BTT SKR 1.4 / 1.4 Turbo

### PR DESCRIPTION
**Pull request by StoneColdCrazy**
### Requirements
Bigtreetech SKR 1.4/1.4 Turbo multiply endstops usage

### Description
Bigtreetech board has defined only 3 endstop pins. If more needed (eg. when dual endstops needed) you have to change the pins directly in the file, because the pins are used for 3 other functions.
The definition of X_STOP_PIN, Y_STOP_PIN and Z_STOP_PIN also prevents to set an these es limit switches and has to be defined as MIN pins. 

### Benefits
When set in Configuration.h per define (ENABLE_XMAX_PIN, ENABLE_YMAX_PIN and/or ENABLE_ZMAX_PIN) the previous definitions disabled and MIN_PIN/MAX_PIN for endtsops are set correctly

### CODE REVIEW

**Comment by thisiskeithb** [comment]
> Bigtreetech board has defined only 3 endstop pins. 

[This is how the board was designed](https://github.com/bigtreetech/BIGTREETECH-SKR-V1.3/blob/master/BTT%20SKR%20V1.4/Hardware/BTT%20SKR%20V1.4PIN.pdf), but remapping the PS/filament runout pins will work if they aren’t being used:

image

**Comment by StoneColdCrazy** [comment]
@thisiskeithb this is correct, but many people using more than an single endstop. If you are coming from a nearly identical (pcb layout) SKR1.1 or SKR1.3 than this is a little bit confusing.

**Comment by thinkyhead** [comment]
I see what you're going for here, but this isn't really necessary. Following the existing config, you are of course free to use all unused pins for your own purposes. Unless the plan is to propagate these "`ENABLE_****_PIN`" options to every board, this change only makes this one pins file quirky.

**Comment by thinkyhead** [comment]
If the board actually has 6 endstop pins on it labeled as X+/X- Y+/Y- Z+/Z- then those pins should be defined as such in the pins file, and not just as STOP pins. If the board has built-in fixed routing of the DIAG pins of Trinamic stepper drivers, then the DIAG pins need to be defined in the pins file and then these will be assigned to the endstops according to the use of stallGuard on each axis.

**Comment by thinkyhead** [comment]
I've applied changes to match the approach first applied for SKR 1.3. When using stallGuard the DIAG pins are assigned to the appropriate endstop and the other endstop is defined to the "other" pin.

Note that —except for the ones actually used as endstops— the endstop pins are not (generally) used by name in the rest of the code, so there's no point in trying to expose or make the "other" pins available by the original approach here, since they will get assigned to other functions anyway.

**Comment by thinkyhead** [comment]
I see that the board can also do Serial UART, so I'm curious whether that can replace the DIAG pin for stallGuard, or if stallGuard still uses the DIAG pin in all cases….

**Comment by thisiskeithb** [comment]
> If the board actually has 6 endstop pins on it labeled as X+/X- Y+/Y- Z+/Z- then those pins should be defined as such in the pins file, and not just as STOP pins.

The SKR 1.4 just has X/Y/Z min pins (See the pins diagram in my comment above), but the PS/filament runout pins are the same connector/spot where the X/Y/Z max endstops were on the SKR 1.3.

**Comment by ManuelMcLure** [comment]
Do we know whether the end stop connectors have hardwired pull-ups? I'm wondering if we should add ``ONBOARD_ENDSTOPPULLUPS`` to the 1.4 pins file. I expect the PS pin to _not_ have an onboard pull-up, but the rest probably do.

**Comment by thinkyhead** [comment]
> The SKR 1.4 just has X/Y/Z min pins

Sorta. Two of the 3 pins next door are routed to the DIAG pins on E0 and E1, so technically they can function as extra endstops when using dual/triple/quad XYZ. Of course E1 is the only one typically used, but if someone wants to use a config with zero extruders, E0 could also be used.

**Comment by thinkyhead** [comment]
> I expect the PS pin to not have an onboard pull-up, but the rest probably do.

I hope for consistency and symmetry that they all do. If the pullups are in the MCU then I expect even more strongly that it's just another pin of the same flavor.

**Comment by thinkyhead** [comment]
I see 16 resistors stacked up next to these ports, but none of the documents in the BTT github repo show the traces on the back very clearly. There is a photo of the back but it's hard to follow. ~~No STEP or EAGLE files that I can see….~~

Aha, I do see a STEP file… checking it out…

**Comment by ManuelMcLure** [comment]
On the SKR 1.3 (and Re-ARM) they're on the board, not the MCU. The SKR 1.3 has more resistors/caps than the SKR 1.4 does. Assuming two resistors and a cap per pin which is what the SKR 1.3 uses, the 16 isn't sufficient for all 6 pins. The SKR 1.3 has 19 caps/resistors near the connectors which would work for all 6 (I don't know what the extra resistor (R22) is for). The reason I suspect that the PS pin _doesn't_ have a pull-up is that it can be used either for power detection or as a ``PS_ON`` pin for power control, and if it had a hardwired pull-up that wouldn't be possible.
**EDIT:** It looks like on the 1.3 R22 is a pullup on P0_28 which is on the EXP1 connector.

**Comment by thinkyhead** [comment]
I'll add `ONBOARD_ENDSTOPPULLUPS` for the sake of the limit switches.

I assume this will also be consistent with the way DIAG pins are pulled…?

Also, am I mistaken to think that with the pins always pulled-up you can only use one flavor of switches — either NC or NO but not both…?

The `PS_ON_PIN` will be treated as usual — as an OUTPUT pin.

**Comment by ManuelMcLure** [comment]
Both NC and NO switches should have a pull-up so the onboard pull-ups should not cause any issues using a mix of NC and NO switches as long as you set ``_ENDSTOP_INVERTING`` appropriately.
However, I believe the pull-ups might cause problems with PNP sensors - luckily most common sensors are meant to replace a switch and are thus NPN which work fine with the pull-ups.

**Comment by thinkyhead** [comment]
> Both NC and NO switches should have a pull-up

I've never looked too closely, but do common leaf endstop switches or optos have embedded pull-ups?

**Comment by ManuelMcLure** [comment]
The standalone micro switches don’t have a pull-up - just a common, NO, and NC connector. However the ones mounted on PCBs often do.
Both opto endstops described on reprap.org have pull-ups.

**Comment by KCGreg** [comment]
I just hooked up one of these boards with X, Y, Z, E0, and Z2 in UART mode with sensorless homing and ABL using a BLTouch 3.1 and TMC 2209 in all sockets.  I used the built in servo and probe connector (.10) for the BLTouch.  In order to get it to work, I had to home to z max as zmin was always showing as tripped due to the built in diag connection. Once I made this change, it homed to max and probed the bed just fine.  I did cut the diag pin on the 2209 used for E0 so that I can use the XMax(E0Det) connector for a filament runout sensor.
